posts/2022/01/ipython-8.0-lessons-learned-maintaining-software.md

```diff
@@ -0,0 +1,303 @@
+Make the codebase faster to navigate and simpler to understand.
+
+As Brian Kernighan put it: “Debugging is twice as hard as writing the code in the first place. Therefore, if you write
+the code as cleverly as possible, you are, by definition, not smart enough to debug it.”
+
+So I'd like to keep my code as simple as possible with less states or condition to keep in mind. 
+
+Seen it in another way, your software progressing both because it's leading edge is moving forward, as it does because
+the trailing end is also catching up. And for this release of IPython we focused on both. 
+
+
+## LBYL vs EAFP
+
+Look Before You Leap and Easier to Ask Forgiveness than Permissions are two practices each with their advantages and
+inconvenient. Python tend the prefer the second one as try/except is "easy" in Python. 
+
+This first implementation will often feel more idiomatic Python:
+
+```python
+from pahtlib import Path
+
+p = Path('test')
+
+try:
+   p.mkdir()
+Except FileExistsError:
+   continue
+
+```
+
+While this second one can feel less Pythonistic
+
+```
+if not p.exists():
+    p.mkdir()
+```
+
+In particular note that the second one suffer a race condition, as the directory could be created between the check and
+call to mkdir and that anyway you should use the `exist_ok=True` argument.
+
+Regardless I have found that conditional code using try and except / EAFP to deal with various versions of a dependency
+is bad practice. It make figuring out the reason for these exception hard to discover, and also make it hard to find all
+occurrences of such conditional imports. I guess that should have bee obvious from the Zen of Python:
+
+```
+In [1]: import this
+The Zen of Python, by Tim Peters
+
+...
+Explicit is better than implicit.
+...
+```
+
+So my first tip: Always avoid catching  `ImportError`s when you can compare version numbers. 
+
+ - having two import lines will not make import slower. 
+ - it's explicit.
+ - it's easier to search for. 
+ - it's easier to remove.
+
+
+ For example IPython used to contain
+
+```
+try:
+    from numpy.testing import KnownFailure, knownfailureif
+except ImportError:
+    from ._decorators import knownfailureif
+    try:
+        from ._numpy_testing_noseclasses import KnownFailure
+    except ImportError:
+        pass
+```
+
+
+Which adds a fallback for [numpy version older than numpy 1.3 from 2008](https://github.com/numpy/numpy/commit/ba9a02dcb2c3ca635076a75cc9eb0f406e00ceed).
+It tooks me ~30 minutes to find this informations, which could have been seconds would the author (which could have been
+me) had checked version.  A proper version check would also had this code removed years ago. 
+
+While IPython has few dependencies beyond Python, we do so with Python version itself, and always compare with
+`sys.version_info` made it straightforward to find all dead code once we bumped or minimal version to 3.8+.
```

Comment:
The above example has to deal with `numpy`. I think readers would be interested in how you are comparing numpy versions now. This can be done either with a link to IPython code or a short code snippet.

Reply:
Good point, I'm not comparing numpy version anymore as we just say in setup.py

```
 test_extra=[
...
        "numpy>=1.19",
...
    ],
```

But I'll extend.